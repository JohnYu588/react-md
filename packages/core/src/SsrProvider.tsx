import type { ReactElement, ReactNode } from "react";
import { createContext, useContext, useEffect, useRef } from "react";
import type { NonNullRef } from "./types";

const context = createContext<NonNullRef<boolean>>({ current: false });
const { Provider } = context;
context.displayName = "Ssr";

/**
 * @internal
 * @remarks \@since 6.0.0
 */
export function useSsr(): boolean {
  return useContext(context).current;
}

/**
 * @internal
 * @remarks \@since 6.0.0
 */
export interface SsrProviderProps {
  ssr?: boolean;
  children: ReactNode;
}

/**
 * @internal
 * @remarks \@since 6.0.0
 */
export function SsrProvider(props: SsrProviderProps): ReactElement {
  const { ssr = false, children } = props;
  const ssrRef = useRef(ssr);
  useEffect(() => {
    ssrRef.current = false;
  }, []);

  return <Provider value={ssrRef}>{children}</Provider>;
}
