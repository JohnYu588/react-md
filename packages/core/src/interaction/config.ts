import { type ElementInteractionMode } from "./types.js";

/**
 * @remarks \@since 6.0.0
 */
export interface ElementInteractionConfiguration {
  mode: ElementInteractionMode;
  higherContrast: boolean;
}

// NOTE: Uses get/set for test mocking

let mode: ElementInteractionMode = "ripple";
let higherContrast = true;

/**
 * @remarks \@since 6.0.0
 */
export const INTERACTION_CONFIG: ElementInteractionConfiguration = {
  get mode() {
    return mode;
  },
  set mode(nextMode: ElementInteractionMode) {
    mode = nextMode;
  },
  get higherContrast() {
    return higherContrast;
  },
  set higherContrast(nextHigherContrast) {
    higherContrast = nextHigherContrast;
  },
};
