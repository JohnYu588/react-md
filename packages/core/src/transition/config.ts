/**
 * @since 6.0.0
 */
export interface TransitionConfig {
  /**
   * Set this to `true` to disable all transitions from `react-md`.
   *
   * Note: It is recommended to set this to `true` in testing to keep things
   * simple and will automatically be set when using:
   *
   * ```ts
   * import "@react-md/core/test-utils/jest-globals/setup";
   *
   * // or
   * import "@react-md/core/test-utils/vitest/setup";
   * ```
   *
   * @defaultValue `false`
   */
  disabled: boolean;
}

// NOTE: Uses get/set for test mocking

let disabled = false;

/**
 * @since 6.0.0
 */
export const TRANSITION_CONFIG: TransitionConfig = {
  get disabled() {
    return disabled;
  },
  set disabled(nextDisabled: boolean) {
    disabled = nextDisabled;
  },
};
