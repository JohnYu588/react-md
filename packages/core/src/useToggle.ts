"use client";

import { useCallback, useState } from "react";

import { type UseStateInitializer, type UseStateSetter } from "./types.js";

/**
 * @since 6.0.0
 */
export interface ToggleImplementation {
  toggled: boolean;
  setToggled: UseStateSetter<boolean>;
  toggle: () => void;
  enable: () => void;
  disable: () => void;
}

/**
 *
 * @example Simple Example
 * ```tsx
 * import { Button, useToggle } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   const { toggled, toggle } = useToggle();
 *
 *   return (
 *     <>
 *       <Button onClick={toggle}>Toggle</Button>
 *       {`Toggled: ${toggled}`}
 *     </>
 *   );
 * }
 * ```
 *
 * @param defaultValue - `false`
 */
export function useToggle(
  defaultValue: UseStateInitializer<boolean> = false
): ToggleImplementation {
  const [toggled, setToggled] = useState(defaultValue);

  return {
    toggled,
    setToggled,
    toggle: useCallback(() => {
      setToggled((prevToggled) => !prevToggled);
    }, []),
    enable: useCallback(() => {
      setToggled(true);
    }, []),
    disable: useCallback(() => {
      setToggled(false);
    }, []),
  };
}
