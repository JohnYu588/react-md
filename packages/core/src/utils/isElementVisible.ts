/**
 * Note: requires the `$disable-display-none-class` variable to remain `false`.
 *
 * @remarks \@since 6.0.0
 */
export const DISPLAY_NONE_CLASS = "rmd-display-none";

/**
 * This is a small util to check if an element within react-md is visible by
 * checking the element and all parents to see if they contain the
 * {@link DISPLAY_NONE_CLASS}.
 *
 * @example
 * Simple Example
 * ```tsx
 * const treeItem = screen.getByRole("treeitem", { name: "Tree Item" });
 * const subTreeItem = screen.getByRole("treeitem", { Name: "Sub Tree Item" });
 *
 * expect(isElementVisible(treeItem)).toBe(true);
 * expect(isElementVisible(subTreeItem)).toBe(false):
 *
 * await user.click(treeItem);
 * expect(isElementVisible(treeItem)).toBe(true);
 * expect(isElementVisible(subTreeItem)).toBe(true);
 * ```
 * @remarks \@since 6.0.0
 */
export function isElementVisible(element: HTMLElement | null): boolean {
  if (!element) {
    return false;
  }

  let currentElement: HTMLElement | null = element;
  while (currentElement) {
    if (currentElement.classList.contains(DISPLAY_NONE_CLASS)) {
      return false;
    }

    currentElement = currentElement.parentElement;
  }

  return true;
}
