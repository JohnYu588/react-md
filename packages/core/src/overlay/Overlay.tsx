"use client";
import { forwardRef, type HTMLAttributes } from "react";
import { type BoxAlignItems, type BoxJustifyContent } from "../box/styles.js";
import { Portal } from "../portal/Portal.js";
import { useSsr } from "../SsrProvider.js";
import {
  type CSSTransitionComponentProps,
  type TransitionActions,
} from "../transition/types.js";
import { useCSSTransition } from "../transition/useCSSTransition.js";
import {
  DEFAULT_OVERLAY_CLASSNAMES,
  DEFAULT_OVERLAY_TIMEOUT,
  overlay,
} from "./overlayStyles.js";

/**
 * @remarks \@since 6.0.0 Added `align` and `justify` props.
 */
export interface OverlayProps
  extends HTMLAttributes<HTMLSpanElement>,
    CSSTransitionComponentProps,
    TransitionActions {
  /**
   * @defaultValue `"center"`
   * @remarks \@since 6.0.0
   */
  align?: BoxAlignItems;

  /**
   * @defaultValue `"center"`
   * @remarks \@since 6.0.0
   */
  justify?: BoxJustifyContent;

  /**
   * Set this to `true` for when the overlay should be visible. Toggling this
   * value will trigger the enter/exit animation.
   */
  visible: boolean;

  /**
   * Set this to `true` if the overlay should be rendered with an `opacity: 0`
   * and disabling the animation. This is useful if you'd like a "close on
   * outside click" behavior.
   *
   * @defaultValue `false`
   */
  noOpacity?: boolean;

  /**
   * @see {@link OverlayClassNameOptions.clickable}
   * @defaultValue `!noOpacity`
   */
  clickable?: boolean;

  /**
   * @defaultValue `false`
   */
  disablePortal?: boolean;

  /**
   * @defaultValue `false`
   */
  disableTransition?: boolean;
}

/**
 * **Client Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Button, Overlay, useToggle } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   const { toggle, disable, toggled: visible } = useToggle(false);
 *
 *   return (
 *     <>
 *       <Button onClick={toggle}>Toggle</Button>
 *       <Overlay visible={visible} onClick={disable} />
 *     </>
 *   );
 * }
 * ```
 *
 * @remarks \@since 6.0.0 Removed the `onRequestClose` prop in favor of using
 * the `onClick` prop instead.
 */
export const Overlay = forwardRef<HTMLSpanElement, OverlayProps>(
  function Overlay(props, nodeRef) {
    const {
      children,
      className,
      visible,
      noOpacity = false,
      clickable = !noOpacity,
      temporary = true,
      timeout = DEFAULT_OVERLAY_TIMEOUT,
      classNames = DEFAULT_OVERLAY_CLASSNAMES,
      disableTransition = false,
      align = "center",
      justify = "center",
      appear,
      enter,
      exit,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      exitedHidden = true,
      disablePortal: propDisablePortal = false,
      ...remaining
    } = props;

    const ssr = useSsr();
    const { elementProps, rendered, disablePortal } = useCSSTransition({
      nodeRef,
      transitionIn: visible,
      timeout: noOpacity ? 0 : timeout,
      classNames: noOpacity ? "" : classNames,
      className: overlay({
        visible,
        clickable,
        align,
        justify,
        className,
      }),
      appear: appear && !disableTransition && !ssr,
      enter: enter && !disableTransition,
      exit: exit && !disableTransition,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      temporary,
      exitedHidden,
      disablePortal: propDisablePortal,
    });

    return (
      <Portal disabled={disablePortal}>
        {rendered && (
          <span {...remaining} {...elementProps}>
            {children}
          </span>
        )}
      </Portal>
    );
  }
);
