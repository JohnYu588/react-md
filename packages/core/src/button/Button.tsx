"use client";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import { RippleContainer } from "../interaction/RippleContainer";
import { useElementInteraction } from "../interaction/useElementInteraction";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren";
import type { PropsWithRef } from "../types";
import { bem } from "../utils";
import type {
  FloatingActionButtonPosition,
  FloatingActionButtonProps,
} from "./FloatingActionButton";
import { FloatingActionButton } from "./FloatingActionButton";

declare module "react" {
  interface CSSProperties {
    "--rmd-button-background-color"?: string;
    "--rmd-button-color"?: string;
    "--rmd-button-border-radius"?: string | number;
    "--rmd-button-contained-background-color"?: string;
    "--rmd-button-contained-color"?: string;
    "--rmd-button-text-horizontal-padding"?: string | number;
    "--rmd-button-text-vertical-padding"?: string | number;
    "--rmd-button-text-min-height"?: string | number;
    "--rmd-button-text-min-width"?: string | number;
    "--rmd-button-icon-size"?: string | number;
    "--rmd-button-icon-radius"?: string | number;
  }
}

const styles = bem("rmd-button");

/**
 * When this is set to `"text"`, the size of the button will be determined by
 * the content and will be more block-like. Icons can still be rendered in
 * `"text"` buttons and will have spacing automatically applied between other
 * content in the button.
 *
 * When this is set to `"icon"`, the button will be equal height/width and
 * circular.
 */
export type ButtonType = "text" | "icon" | "icon-square";

/**
 * One of the valid material design default button themes that can be used. This
 * will update the general look and feel by updating the colors within the
 * button while the `ButtonThemeType` will update the borders or box shadow.
 */
export type ButtonTheme =
  | "clear"
  | "primary"
  | "secondary"
  | "warning"
  | "error"
  | "disabled";

/**
 * When this is set to `"flat"`, the button will have no `background-color`,
 * `border`, and `box-shadow`. It will only set the `color` to the
 * {@link ButtonTheme}.
 *
 * When this is set to `"outline"`, the button will have no `background-color`,
 * but gain a `border` and `color` set to the {@link ButtonTheme}.
 *
 * When this is set to `"contained"`, the button will set the `background-color`
 * to the {@link ButtonTheme}, add some `box-shadow`, and set the `color` to
 * either `#000` or `#fff`. (The `color` defaults to whichever value has the
 * highest contrast ratio with the `background-color`)
 */
export type ButtonThemeType = "flat" | "outline" | "contained";

/** @remarks \@since 6.0.0 */
export interface ButtonClassNameThemeOptions {
  className?: string;

  /** @defaultValue `false` */
  disabled?: boolean;

  /**
   * @see {@link ButtonTheme} for information about the different types.
   * @defaultValue `"text"`
   */
  buttonType?: ButtonType;

  /** @defaultValue `"clear"` */
  theme?: ButtonTheme;

  /**
   * @see {@link ButtonThemeType} for information about the theming behavior.
   * @defaultValue `"flat"`
   */
  themeType?: ButtonThemeType;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ButtonClassNameOptions extends ButtonClassNameThemeOptions {
  /** @defaultValue `false` */
  pressed?: boolean;
  pressedClassName?: string;
}

/**
 * Creates a button theme based on the button theming props. This is really just
 * used so that other elements like clickable `<div>`s or `<input type="file">`
 * can look like buttons.
 *
 * @param options - An object containing the themeable button props to generate a
 * button theme className.
 * @returns a string of class names to create an element with a button theme.
 * @remarks \@since 6.0.0 This used to be called `buttonThemeClassNames`.
 */
export function button(options: ButtonClassNameOptions = {}): string {
  const {
    theme: propTheme = "clear",
    themeType = "flat",
    buttonType = "text",
    disabled: propDisabled = false,
    pressed = false,
    pressedClassName,
    className,
  } = options;

  const theme = propTheme === "disabled" ? "clear" : propTheme;
  const disabled = propDisabled || propTheme === "disabled";
  const text = buttonType === "text";
  const icon = !text;
  const outline = themeType === "outline";
  const contained = themeType === "contained";
  const clear = theme === "clear";

  return cnb(
    styles({
      text,
      icon,
      "icon-square": buttonType === "icon-square",
      disabled,
      contained: !disabled && contained,
      outline,
      pressed: contained && pressed,
      [theme]: !disabled && !clear && contained,
      [`text-${theme}`]: !disabled && !clear && !contained,
      [`outline-${theme}`]: !disabled && !clear && outline,
    }),
    pressedClassName,
    className
  );
}

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";

  /**
   * The position within the viewport to display the button as a floating action
   * button.
   */
  floating?: FloatingActionButtonPosition;

  /**
   * Any additional props to provide the to `FAB` container element when the
   * `floating` prop is provided
   */
  floatingProps?: PropsWithRef<FloatingActionButtonProps, HTMLSpanElement>;
}

/**
 * **Client Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      floating = null,
      floatingProps,
      theme = floating ? "secondary" : "clear",
      themeType = floating ? "contained" : "flat",
      buttonType = floating ? "icon" : "text",
      className,
      children: propChildren,
      onBlur,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onDragStart,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onBlur,
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onDragStart,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <FloatingActionButton position={floating} {...floatingProps}>
        <button
          {...remaining}
          // when the theme is set to `"disabled"`, the event handlers should be
          // removed so that it behaves like a disabled button. you do not want to
          // actually set the `disabled` attribute since it will lose keyboard
          // focus. this is mostly for supporting circular progress bars within
          // buttons
          {...(isThemeDisabled ? {} : handlers)}
          aria-disabled={isThemeDisabled || remaining["aria-disabled"]}
          disabled={disabled}
          ref={ref}
          type={type}
          className={button({
            theme,
            themeType,
            buttonType,
            disabled,
            pressed,
            pressedClassName,
            className,
          })}
        >
          {children}
          {rippleContainerProps && (
            <RippleContainer {...rippleContainerProps} />
          )}
        </button>
      </FloatingActionButton>
    );
  }
);
