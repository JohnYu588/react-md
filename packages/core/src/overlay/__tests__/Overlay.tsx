import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { createRef, type ReactElement } from "react";
import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from "../../test-utils/index.js";

import { SsrProvider } from "../../SsrProvider.js";
import { Button } from "../../button/Button.js";
import { TRANSITION_CONFIG } from "../../transition/config.js";
import { useToggle } from "../../useToggle.js";
import { DISPLAY_NONE_CLASS } from "../../utils/isElementVisible.js";
import { Overlay, type OverlayProps } from "../Overlay.js";

interface TestProps {
  defaultVisible?: boolean;
}

function Test({ defaultVisible = false }: TestProps): ReactElement {
  const { toggle, disable, toggled: visible } = useToggle(defaultVisible);

  return (
    <>
      <Button onClick={toggle}>Toggle</Button>
      <Overlay data-testid="overlay" visible={visible} onClick={disable} />
    </>
  );
}

describe("Overlay", () => {
  beforeEach(() => {
    TRANSITION_CONFIG.disabled = false;
  });

  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLSpanElement>();
    const props = {
      "data-testid": "overlay",
      ref,
      visible: true,
    } as const;

    const { rerender } = render(<Overlay {...props} />);

    const element = screen.getByTestId("overlay");
    expect(ref.current).toBeInstanceOf(HTMLSpanElement);
    expect(ref.current).toBe(element);
    expect(element).toMatchSnapshot();

    rerender(
      <Overlay
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(element).toMatchSnapshot();
  });

  it("should support rendering inline if disablePortal prop is true", () => {
    const { container } = render(
      <Overlay data-testid="overlay" visible disablePortal />
    );

    const overlay = screen.getByTestId("overlay");
    expect(container.firstElementChild).toBe(overlay);
    expect(container).toMatchSnapshot();
  });

  it("should support rendering inline if it was visible by default and rendered through ssr", async () => {
    const { container } = render(<Test defaultVisible />, {
      wrapper: ({ children }) => <SsrProvider ssr>{children}</SsrProvider>,
    });

    const button = screen.getByRole("button", { name: "Toggle" });
    const overlay = screen.getByTestId("overlay");
    expect(button.nextElementSibling).toBe(overlay);
    expect(container).toMatchSnapshot();

    fireEvent.click(overlay);
    await waitFor(() => {
      expect(overlay).not.toBeInTheDocument();
    });

    fireEvent.click(button);
    await waitFor(() => {
      expect(() => screen.getByTestId("overlay")).not.toThrow();
    });
    await waitFor(() => {
      expect(() => within(container).getByTestId("overlay")).toThrow();
    });
  });

  it("should allow for overriding CSS transition options", async () => {
    const onEnter = jest.fn();
    const onEntering = jest.fn();
    const onEntered = jest.fn();
    const onExit = jest.fn();
    const onExiting = jest.fn();
    const onExited = jest.fn();

    const props: Partial<OverlayProps> & { "data-testid": string } = {
      "data-testid": "overlay",
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      temporary: false,
    };

    const { rerender } = render(<Overlay {...props} visible={false} />);

    const overlay = screen.getByTestId("overlay");
    expect(overlay).toHaveClass(DISPLAY_NONE_CLASS);
    expect(overlay).toMatchSnapshot();

    expect(onEnter).not.toHaveBeenCalled();
    expect(onEntering).not.toHaveBeenCalled();
    expect(onEntered).not.toHaveBeenCalled();
    expect(onExit).not.toHaveBeenCalled();
    expect(onExiting).not.toHaveBeenCalled();
    expect(onExited).not.toHaveBeenCalled();

    rerender(<Overlay {...props} visible />);
    await waitFor(() => {
      expect(onEntered).toHaveBeenCalledTimes(1);
    });
    expect(onEnter).toHaveBeenCalledTimes(1);
    expect(onEntering).toHaveBeenCalledTimes(1);
    expect(onExit).not.toHaveBeenCalled();
    expect(onExiting).not.toHaveBeenCalled();
    expect(onExited).not.toHaveBeenCalled();
    expect(overlay).not.toHaveClass(DISPLAY_NONE_CLASS);
    expect(overlay).toMatchSnapshot();

    rerender(<Overlay {...props} visible={false} />);
    await waitFor(() => {
      expect(onExited).toHaveBeenCalledTimes(1);
    });
    expect(onEnter).toHaveBeenCalledTimes(1);
    expect(onEntering).toHaveBeenCalledTimes(1);
    expect(onEntered).toHaveBeenCalledTimes(1);
    expect(onExit).toHaveBeenCalledTimes(1);
    expect(onExiting).toHaveBeenCalledTimes(1);
    expect(overlay).toHaveClass(DISPLAY_NONE_CLASS);
    expect(overlay).toMatchSnapshot();
  });

  it("should disable the transition if the noOpacity prop is true", async () => {
    const onEnter = jest.fn();
    const onEntering = jest.fn();
    const onEntered = jest.fn();
    const onExit = jest.fn();
    const onExiting = jest.fn();
    const onExited = jest.fn();
    const props = {
      "data-testid": "overlay",
      visible: false,
      noOpacity: true,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
    } as const;

    const { rerender } = render(<Overlay {...props} />);
    expect(() => screen.getByTestId("overlay")).toThrow();

    rerender(<Overlay {...props} visible />);
    await waitFor(() => {
      expect(onEntered).toHaveBeenCalledTimes(1);
    });
    expect(onEnter).not.toHaveBeenCalled();
    expect(onEntering).not.toHaveBeenCalled();
    expect(onExit).not.toHaveBeenCalled();
    expect(onExiting).not.toHaveBeenCalled();
    expect(onExited).not.toHaveBeenCalled();

    const overlay = screen.getByTestId("overlay");
    expect(overlay).toMatchSnapshot();
  });

  it("should support passing the box align/justify to the box class name utils", () => {
    render(
      <Overlay visible align="start" justify="end" data-testid="overlay" />
    );
    const overlay = screen.getByTestId("overlay");

    expect(overlay).toMatchSnapshot();
  });
});
