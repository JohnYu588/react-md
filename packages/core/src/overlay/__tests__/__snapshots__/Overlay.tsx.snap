// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay should allow for overriding CSS transition options 1`] = `
<span
  class="rmd-overlay rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center rmd-display-none"
  data-testid="overlay"
/>
`;

exports[`Overlay should allow for overriding CSS transition options 2`] = `
<span
  class="rmd-overlay rmd-overlay--visible rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center rmd-overlay--active"
  data-testid="overlay"
/>
`;

exports[`Overlay should allow for overriding CSS transition options 3`] = `
<span
  class="rmd-overlay rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center rmd-display-none"
  data-testid="overlay"
/>
`;

exports[`Overlay should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<span
  class="rmd-overlay rmd-overlay--visible rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center rmd-overlay--active"
  data-testid="overlay"
/>
`;

exports[`Overlay should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<span
  class="rmd-overlay rmd-overlay--visible rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center custom-class-name rmd-overlay--active"
  data-testid="overlay"
  style="color: white;"
/>
`;

exports[`Overlay should disable the transition if the noOpacity prop is true 1`] = `
<span
  class="rmd-overlay rmd-overlay--visible rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center"
  data-testid="overlay"
/>
`;

exports[`Overlay should support passing the box align/justify to the box class name utils 1`] = `
<span
  class="rmd-overlay rmd-overlay--visible rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-start rmd-box--justify-end rmd-overlay--active"
  data-testid="overlay"
/>
`;

exports[`Overlay should support rendering inline if disablePortal prop is true 1`] = `
<div>
  <span
    class="rmd-overlay rmd-overlay--visible rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center rmd-overlay--active"
    data-testid="overlay"
  />
</div>
`;

exports[`Overlay should support rendering inline if it was visible by default and rendered through ssr 1`] = `
<div>
  <button
    class="rmd-button rmd-button--text rmd-interaction-surface"
    type="button"
  >
    <span>
      Toggle
    </span>
  </button>
  <span
    class="rmd-overlay rmd-overlay--visible rmd-overlay--clickable rmd-box rmd-box--gap rmd-box--wrap rmd-box--align-center rmd-box--justify-center rmd-overlay--active"
    data-testid="overlay"
  />
</div>
`;
